### PR TITLE
Unifiedpush

### PR DESCRIPTION
**Pull request by p1gp1g**
### Pull Request Checklist

- [x] Changes has been tested on an Android device or Android emulator with API 21
- [x] UI change has been tested on both light and dark themes
- [x] Pull request is based on the develop branch
- [x] Pull request updates [CHANGES.md](https://github.com/vector-im/element-android/blob/develop/CHANGES.md)
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)

It still need some improvement, and help is welcomed.
- [ ] Test the gateway and use the user's one if present.¹
- [ ] Simplify diff Fdroid/Gplay.²
- [ ] Maybe some improvement on the checking list.

¹ im.vector.app.core.pushers.UPHelper.customOrDefaultGateway: GET custom, if it returns '{"gateway":"matrix"} use custom. What is the best way to do the GET on element-android point of view?
² the only diff should be im.vector.app.push.fcm.FCMHelper.isPlayServicesAvailable and im.vector.app.push.fcm.EmbeddedDistrib. And so, the fallback should work on all flavours without a distributor (on Gplay version, it happens only if the gplay version has been installed on a phone without GServices).

### CODE REVIEW

File: vector/src/main/java/im/vector/app/core/pushers/VectorMessagingReceiver.kt [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            data = JSONObject(message)
+            notification = data.getJSONObject("notification")
+        } catch (e: JSONException) {
+            e.printStackTrace()
+            return
+        }
+        try {
+            eventId = notification.getString("event_id")
+        } catch (e: JSONException) {
+            Timber.i("No event_id on notification")
```
Comment (ShadowJonathan) [inline vector/src/main/java/im/vector/app/core/pushers/VectorMessagingReceiver.kt:78]: Maybe make this (and the other one) `Timber.w`, warnings?

File: vector/src/main/java/im/vector/app/core/pushers/VectorMessagingReceiver.kt [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        initVar(context!!)
+        Timber.i("onMessage received")
+        lateinit var data: JSONObject
+        lateinit var notification: JSONObject
+        lateinit var eventId: String
+        try {
+            data = JSONObject(message)
+            notification = data.getJSONObject("notification")
+        } catch (e: JSONException) {
+            e.printStackTrace()
```
Comment (ShadowJonathan) [inline vector/src/main/java/im/vector/app/core/pushers/VectorMessagingReceiver.kt:72]: How about also a log message what went wrong? `Timber.e`

File: vector/src/main/java/im/vector/app/core/pushers/VectorMessagingReceiver.kt [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            notification.put("room_id", "")
+        }
+        if (eventId == PushersManager.TEST_EVENT_ID) {
+            val intent = Intent(NotificationUtils.PUSH_ACTION)
+            LocalBroadcastManager.getInstance(context).sendBroadcast(intent)
+            return
+        }
+
+        if (!vectorPreferences.areNotificationEnabledForDevice()) {
+            Timber.i("Notification are disabled for this device")
```
Comment (ShadowJonathan) [inline vector/src/main/java/im/vector/app/core/pushers/VectorMessagingReceiver.kt:95]:
```suggestion
            Timber.i("Notification are disabled for this device, ignoring push event...")
```

File: vector/src/main/java/im/vector/app/core/pushers/VectorMessagingReceiver.kt [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        if (vectorPreferences.areNotificationEnabledForDevice() && activeSessionHolder.hasActiveSession()) {
+            val gateway = UPHelper.customOrDefaultGateway(context, endpoint)
+            UPHelper.storePushGateway(context, gateway)
+            UPHelper.storeUpEndpoint(context, endpoint)
+            pusherManager.registerPusher(endpoint, gateway)
+        }
+    }
+
+    override fun onRegistrationFailed(context: Context?, instance: String) {
+        Toast.makeText(context, "Registration Failed", Toast.LENGTH_SHORT).show()
```
Comment (ShadowJonathan) [inline vector/src/main/java/im/vector/app/core/pushers/VectorMessagingReceiver.kt:122]:
```suggestion
        Toast.makeText(context, "Push service registration failed", Toast.LENGTH_SHORT).show()
```

File: vector/src/main/java/im/vector/app/core/pushers/VectorMessagingReceiver.kt [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    override fun onRegistrationRefused(context: Context?, instance: String) {
+        Toast.makeText(context, "Registration Refused", Toast.LENGTH_SHORT).show()
```
Comment (ShadowJonathan) [inline vector/src/main/java/im/vector/app/core/pushers/VectorMessagingReceiver.kt:126]:
```suggestion
        Toast.makeText(context, "Push service registration refused by server", Toast.LENGTH_LONG).show()
```

(also check if `LENGTH_LONG` actually exists)

**Comment by p1gp1g** [comment]
At this moment, the FCM gateway used for the embedded fcm distributor is https://matrix.org/_matrix/push/v1/notify. It would need to work as a gateway for unifiedpush (https://unifiedpush.org/developers/embedded_fcm/) (see https://github.com/vector-im/element-android/pull/2993/files#diff-681374630b8521a655056eecffcdf2b05b6ae12e04c6fb4f08a206b3dd19e983R27)

Also, the default endpoint can be changed if you prefer to host it (https://github.com/vector-im/element-android/pull/2993/files#diff-63b96f47c62869ba88e77317ff6a0b31483a6ebee57f5669c8be0c0eefce5c4aR19)

File: vector/src/main/java/im/vector/app/core/pushers/VectorMessagingReceiver.kt (hunk 1 above)
Comment (p1gp1g) [reply vector/src/main/java/im/vector/app/core/pushers/VectorMessagingReceiver.kt:78]: Notifications without event_id happen when you read a message to decrease the unread number. It happen quite often and isn't a problem

**Comment by p1gp1g** [comment]
AlertDialogs are forbidden ?
```
Search for forbidden patterns in code...
❌ Error: 'android\.app\.AlertDialog' detected 1 time, in file './vector/src/main/java/im/vector/app/core/pushers/UPHelper.kt', at line 18.
Rule: Use AlertDialog form v7 compat lib
1 forbidden pattern detected
```

**Comment by ShadowJonathan** [comment]
> AlertDialogs are forbidden ?

They're not forbidden, you need to use a compatibility library instead of instantiating them directly via that class path

**Comment by p1gp1g** [comment]
The build doesn't work after the ktlint clean. It shouldn't be related to this PR, since I haven't touch it.
`Could not find org.matrix.gitlab.matrix-org:olm:3.2.2.` which is here : https://jitpack.io/org/matrix/gitlab/matrix-org/olm/3.2.2/olm-3.2.2.pom

**Pull request closed without merging**